File: python/src/kalourmade_arkesel/otp.py
import json

from .error_mapping import raise_for_status
from .http import Transport
from .models import OtpGenerateResult, OtpVerifyResult


class OtpResource:
    def __init__(self, transport: Transport):
        self._transport = transport

    def generate(
        self,
        expiry: int,
        length: int,
        medium: str,
        message: str,
        number: str,
        sender_id: str,
        type: str,
    ) -> OtpGenerateResult:
        body = {
            "expiry": expiry,
            "length": length,
            "medium": medium,
            "message": message,
            "number": number,
            "sender_id": sender_id,
            "type": type,
        }
        status, raw = self._transport.request("POST", "/api/otp/generate", json_body=body)
        raise_for_status(status, raw)
        return OtpGenerateResult.from_dict(json.loads(raw))

    def verify(self, code: str, number: str) -> OtpVerifyResult:
        status, raw = self._transport.request(
            "POST", "/api/otp/verify", json_body={"code": code, "number": number}
        )
        raise_for_status(status, raw)
        return OtpVerifyResult.from_dict(json.loads(raw))

File: node/src/otp.ts
import type { Transport } from './http.js';
import { throwIfError } from './errorMapping.js';
import {
  OtpGenerateResult,
  OtpVerifyResult,
  otpGenerateResultFromJson,
  otpVerifyResultFromJson,
} from './models.js';

export interface GenerateOtpParams {
  expiry: number;
  length: number;
  medium: 'sms' | 'voice';
  message: string;
  number: string;
  senderId: string;
  type: 'numeric' | 'alphanumeric';
}

export class OtpResource {
  constructor(private readonly transport: Transport) {}

  async generate(params: GenerateOtpParams): Promise<OtpGenerateResult> {
    const { status, body } = await this.transport.request('POST', '/api/otp/generate', undefined, {
      expiry: params.expiry,
      length: params.length,
      medium: params.medium,
      message: params.message,
      number: params.number,
      sender_id: params.senderId,
      type: params.type,
    });
    throwIfError(status, body);
    return otpGenerateResultFromJson(JSON.parse(body));
  }

  async verify(params: { code: string; number: string }): Promise<OtpVerifyResult> {
    const { status, body } = await this.transport.request('POST', '/api/otp/verify', undefined, {
      code: params.code,
      number: params.number,
    });
    throwIfError(status, body);
    return otpVerifyResultFromJson(JSON.parse(body));
  }
}
